src/rgw/driver/rados/sync_fairness.cc

```diff
@@ -0,0 +1,354 @@
+    return 0;
+  }
+
+  int restart()
+  {
+    int r = ref.pool.ioctx().unwatch2(handle);
+    if (r < 0) {
+      ldpp_dout(dpp, -1) << "Failed to unwatch on " << ref.obj
+          << " with " << cpp_strerror(-r) << dendl;
+    }
+    r = ref.pool.ioctx().watch2(ref.obj.oid, &handle, this);
+    if (r < 0) {
+      ldpp_dout(dpp, -1) << "Failed to restart watch on " << ref.obj
+          << " with " << cpp_strerror(-r) << dendl;
+      ref.pool.ioctx().close();
+    }
+    return r;
+  }
+
+  void stop()
+  {
+    if (handle) {
+      ref.pool.ioctx().unwatch2(handle);
+      ref.pool.ioctx().close();
+    }
+  }
+
+  // respond to bid notifications
+  void handle_notify(uint64_t notify_id, uint64_t cookie,
+                     uint64_t notifier_id, bufferlist& bl)
+  {
+    if (cookie != handle) {
+      return;
+    }
+
+    BidRequest request;
+    try {
+      auto p = bl.cbegin();
+      decode(request, p);
+    } catch (const buffer::error& e) {
+      ldpp_dout(dpp, -1) << "Failed to decode notification: " << e.what() << dendl;
+      return;
+    }
+
+    BidResponse response;
+    server->on_peer_bid(notifier_id, std::move(request.bids), response.bids);
+
+    bufferlist reply;
+    encode(response, reply);
+
+    ref.pool.ioctx().notify_ack(ref.obj.oid, notify_id, cookie, reply);
+  }
+
+  // reestablish the watch if it gets disconnected
+  void handle_error(uint64_t cookie, int err)
+  {
+    if (cookie != handle) {
+      return;
+    }
+    if (err == -ENOTCONN) {
+      ldpp_dout(dpp, 4) << "Disconnected watch on " << ref.obj << dendl;
+      restart();
+    }
+  }
+}; // Watcher
+
+
+class RadosBidManager;
+
+// RGWRadosNotifyCR wrapper coroutine
+class NotifyCR : public RGWCoroutine {
+  rgw::sal::RadosStore* store;
+  RadosBidManager* mgr;
+  rgw_raw_obj obj;
+  bufferlist request;
+  bufferlist response;
+ public:
+  NotifyCR(rgw::sal::RadosStore* store, RadosBidManager* mgr,
+           const rgw_raw_obj& obj, const bid_vector& my_bids)
+      : RGWCoroutine(store->ctx()), store(store), mgr(mgr), obj(obj)
+  {
+    encode_notify_request(my_bids, request);
+  }
+
+  int operate(const DoutPrefixProvider* dpp) override;
+};
+
+
+class RadosBidManager : public BidManager, public Server, public DoutPrefix {
+  sal::RadosStore* store;
+  rgw_raw_obj obj;
+  Watcher watcher;
+
+  std::mutex mutex;
+  bid_vector my_bids;
+  bidder_map all_bids;
+
+ public:
+  RadosBidManager(sal::RadosStore* store, const rgw_raw_obj& watch_obj,
+                 std::size_t num_shards)
+    : DoutPrefix(store->ctx(), dout_subsys, "sync fairness: "),
+      store(store), obj(watch_obj), watcher(this, store, watch_obj, this)
+  {
+    // fill my_bids with random values
+    std::random_device rd;
+    std::default_random_engine rng{rd()};
+
+    my_bids.resize(num_shards);
+    for(bid_value i = 0; i < num_shards; ++i) {
+      my_bids[i] = i;
+    }
+    std::shuffle(my_bids.begin(), my_bids.end(), rng);
```

Comment:
@ivancich maybe you can help weigh in on this part?

i remember this random shuffle of the integers from `[0, num_shards)` being part of your original design for bid values

but in https://github.com/ceph/ceph/pull/49262, i had instead generated bids using a uniform distribution over the entire domain of `uint16_t`:
```c++
    std::uniform_int_distribution<bid_value> dist;
    my_bids.resize(num_shards);
    std::generate(my_bids.begin(), my_bids.end(), [&] { return dist(rng); });
```
my intuition was that it would be far less likely to result in conflicts, where multiple radosgws generate the same bid for a given shard. if we could avoid these conflicts, that would help minimize contention on these locks

on the other hand, with a relatively small sample size of 64 shards for metadata sync, and 128 shards for data sync, it's likely that the uniform distribution would introduce some bias between the gateways; some would have a higher mean bid value than others

can you think of any middle ground here that might reduce the chance of conflicts without sacrificing fairness?

if not, this shuffle is probably good enough - with the watch-notify approach, lock contention should already be very low

Reply:
Uniform distribution over a smaller range may be?